jxl/src/features/spline.rs

```diff
@@ -270,11 +350,207 @@ pub struct Splines {
     segment_y_start: Vec<usize>,
 }
 
+struct ExtendedPoints<'a> {
+    first_element: Point,
+    last_element: Point,
+    backend: &'a Vec<Point>,
+}
+
+impl ExtendedPoints<'_> {
+    fn len(&self) -> usize {
+        self.backend.len() + 2
+    }
+    fn iter(&self) -> ExtendedPointsIterator {
+        ExtendedPointsIterator {
+            index: 0,
+            backend: self,
+        }
+    }
+}
+
+impl Index<usize> for ExtendedPoints<'_> {
+    type Output = Point;
+    fn index(&self, index: usize) -> &Self::Output {
+        if index == 0 {
+            &self.first_element
+        } else if index == self.len() - 1 {
+            &self.last_element
+        } else {
+            &self.backend[index - 1]
+        }
+    }
+}
+
+struct ExtendedPointsIterator<'a> {
+    index: usize,
+    backend: &'a ExtendedPoints<'a>,
+}
+
+impl Iterator for ExtendedPointsIterator<'_> {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.backend.len() {
+            let result = Some(self.backend[self.index]);
+            self.index += 1;
+            result
+        } else {
+            None
+        }
+    }
+}
+
+fn draw_centripetal_catmull_rom_spline(points: &[Point]) -> Result<Vec<Point>> {
+    if points.is_empty() {
+        return Ok([].to_vec());
+    }
+    if points.len() == 1 {
+        return Ok([points[0]].to_vec());
+    }
+    const NUM_POINTS: usize = 16;
+    // Create a view of points with one prepended and one appended point.
+    let first_point = points[0] + (points[0] - points[1]);
+    let last_point =
+        points[points.len() - 1] + (points[points.len() - 1] - points[points.len() - 2]);
+    let extended_points = iter::once(&first_point)
+        .chain(points.iter())
```

Comment:
Might make sense to do `points.iter().cloned()` and iterate over `Point` instead of `&Point`